Inferred schema type constraints + improved constraints management (#17)
neo4j/docs-spark: modules/ROOT/pages/writing.adoc

work only if `UNIQUE` or `NODE KEY` constraints are defined in Neo4j for the given properties.

[NOTE]

> ===

Reviewer: ```suggestion
====
```

Reviewer: Otherwise, this happens:

<img width="864" alt="Screenshot 2024-02-08 at 11 42 18" src="https://github.com/neo4j/docs-spark/assets/1252064/6d210d6d-3064-4d67-9757-40608729ec7b">

The SaveMode examples apply to the Scala class `org.apache.spark.sql.SaveMode`.  For PySpark,
use a static string with the name of the SaveMode. So instead of `SaveMode.Overwrite`, use `Overwrite` for PySpark.

> ===

Reviewer: ```suggestion
====
```

=== Schema optimization operations

The Spark Connector supports schema optimization operations via:

> * index (deprecated);

Reviewer: ```suggestion
* indexes (deprecated);
```

* constraints;
* property type enforcement;
* set of schema queries.

To speed up the import itself that is executed *before* the import process starts.

[NOTE]
since the version **5.3.0** option **schema.optimization.type** is deprecated in favor of **`schema.optimization.node.\*`**, **`schema.optimization.relationship.*`** and **schema.type.constraint**.

> ==== Deprecated approach

Reviewer: ```suggestion
==== With indexes (Deprecated)
```

Author: Mmm I disagree about this. It's not only related to indexes, I would rather say:
```suggestion
==== `schema.optimization.type` (Deprecated)
```
I think is a best fit

Author: No because it describes both indexes and constraints.
`schema.optimization.type` allows to create indexes and constraints deprecating this approach we do not allows to create indexes anymore but only constraints as is the goal of the new feature.

Reviewer: Ok, I understand now. Sounds good then.

=== Schema optimization operations

The Spark Connector supports schema optimization operations via:

* indexes (deprecated);
* constraints;
* property type enforcement;
* set of schema queries.

To speed up the import itself that is executed *before* the import process starts.

[NOTE]

> since the version **5.3.0** option **schema.optimization.type** is deprecated in favor of **`schema.optimization.node.keys`**, **`schema.optimization.relationship.keys`** and **schema.type.constraint**.

Reviewer: [NOTE]
====
Starting from version **5.3.0**, the `schema.optimization.type` option is deprecated in favor of `schema.optimization.node.keys`, `schema.optimization.relationship.keys`, and `schema.type.constraint`.
====

Author: What I Have to do about this?

Author: Oh I see

|`schema.optimization.node.key` |true/false |false | Create the `NODE KEY` constraint for the properties defined in the option `node.keys`
|`schema.optimization.relationship.unique` |true/false |false | Create the `UNIQUE` constraint for the properties defined in the option `relationship.keys`
|`schema.optimization.relationship.key` |true/false |false | Create the `RELATIONSHIP KEY` constraint for the properties defined in the option `relationship.keys`
|`schema.type.constraint` |true/false |false | Creates the type constraints for node/relationships enforcing the type from the DataFrame schema
|===

===== Node constraints

====== Using `schema.optimization.node.unique`

Given the following example:

[source, scala]
----
    ds.write
      .format(classOf[DataSource].getName)
      .mode(SaveMode.Overwrite)
      .option("url", SparkConnectorScalaSuiteIT.server.getBoltUrl)
      .option("labels", ":Person:Customer")
      .option("node.keys", "surname")
      .option("schema.optimization.node.unique", true)
      .save()
----

Under the hood the Spark connector will create the following constraint:

> `CREATE CONSTRAINT `spark_NODE_UNIQUE-CONSTRAINT_Person_surname` IF NOT EXISTS FOR (e:Person) REQUIRE (e.surname) IS UNIQUE`

Reviewer 2: might be worth calling out that the first label is selected (it is explicitly mentioned in the deprecated schema section)

Reviewer 2: applies to all sections below

Author: Done

====== Using `schema.optimization.relationship.key`

Given the following example:

[source, scala]
----
    ds
      .write
      .mode(SaveMode.Overwrite)
      .format(classOf[DataSource].getName)
      .option("url", SparkConnectorScalaSuiteIT.server.getBoltUrl)
      .option("relationship", "MY_REL")
      .option("relationship.save.strategy", "keys")
      .option("relationship.source.labels", ":NodeA")
      .option("relationship.source.save.mode", "Overwrite")
      .option("relationship.source.node.keys", "idSource:id")
      .option("relationship.target.labels", ":NodeB")
      .option("relationship.target.node.keys", "idTarget:id")
      .option("relationship.target.save.mode", "Overwrite")
      .option("schema.optimization.relationship.key", true)
      .option("relationship.keys", "foo,bar")
      .save()
----

Under the hood the Spark connector will create the following constraint:

`CREATE CONSTRAINT `spark_RELATIONSHIP_KEY-CONSTRAINT_MY_REL_foo-bar` IF NOT EXISTS FOR ()-[e:MY_REL]->() REQUIRE (e.foo, e.bar) IS RELATIONSHIP KEY`

===== Property type constraints

> Since Neo4j 5.11 the database allow to create type constraints for node and relationship properties.

Reviewer 2: ```suggestion
Since Neo4j 5.11 the database allows to create type constraints for node and relationship properties.
```

Author: Done

> In oder to leverage this feature we added the option `schema.type.constraint` that will use the DataFrame schema in order to enforce the type.

Reviewer 2: ```suggestion
In order to leverage this feature we added the option `schema.type.constraint` that will use the DataFrame schema in order to enforce the type.
```

Author: Done

|IntegerType |INTEGER
|LongType |INTEGER
|FloatType |FLOAT
|DoubleType |FLOAT
|DateType |DATE
|TimestampType |LOCAL DATETIME
|Custom `pointType` as: Struct { type: string, srid: integer, x: double, y: double, z: double }| POINT
|Custom `durationType` as: Struct { type: string, months: long, days: long, seconds: long, nanonseconds: integer, value: string }| DURATION
|DataTypes.createArrayType(BooleanType, false) |LIST<BOOLEAN NOT NULL>
|DataTypes.createArrayType(StringType, false) |LIST<STRING NOT NULL>
|DataTypes.createArrayType(IntegerType, false) |LIST<INTEGER NOT NULL>
|DataTypes.createArrayType(LongType, false) |LIST<INTEGER NOT NULL>
|DataTypes.createArrayType(FloatType, false) |LIST<FLOAT NOT NULL>
|DataTypes.createArrayType(DoubleType, false) |LIST<FLOAT NOT NULL>
|DataTypes.createArrayType(DateType, false) |LIST<DATE NOT NULL>
|DataTypes.createArrayType(TimestampType, false) |LIST<LOCAL DATETIME NOT NULL>
|DataTypes.createArrayType(pointType, false) |LIST<POINT NOT NULL>
|DataTypes.createArrayType(durationType, false) |LIST<DURATION NOT NULL>

|===

For the arrays in particular we use the version without null elements as Neo4j does not allow to have them in arrays.

> ====== Node Property type constraints

Reviewer 2: ```suggestion
====== Node Property type and existence constraints
```

Author: Done

Given the following example:

[source, scala]
----
    ds.write
      .format(classOf[DataSource].getName)
      .mode(SaveMode.Overwrite)
      .option("url", SparkConnectorScalaSuiteIT.server.getBoltUrl)
      .option("labels", ":Person:Customer")
      .option("node.keys", "surname")

>       .option("schema.type.constraint", true)

Reviewer 2: Do we expect more settings under `schema.type`? It could just be `schema.type-constraint`, or?
Also, why not `schema.optimization.type.constraint` (or `schema.optimization.type-constraint`) to be more aligned with the others?

Author: Can you please check the document? I remember that was something suggested by @ali-ince (but I'm not sure)
Btw `schema.optimization.type.constraint` could be confused with the old `schema.optimization.type`.

      .save()
----

The connector will create, for each dataframe column a type constraint for the label `Person` according with the mapping table provided above.

The constraint query looks like the following:

[source, cypher]
----
CREATE CONSTRAINT `spark_NODE-TYPE-CONSTRAINT-Person-surname` IF NOT EXISTS FOR (e:Person) REQUIRE e.surname IS :: STRING
----

If the DataFrame schema says that the field is also not null the connector creates an existence constraint as it follows:

[source, cypher]
----
CREATE CONSTRAINT `spark_NODE-NOT_NULL-CONSTRAINT-Person-surname` IF NOT EXISTS FOR (e:Person) REQUIRE e.surname IS NOT NULL
----

> ====== Relationship Property type constraints

Reviewer 2: ```suggestion
====== Relationship Property type and existence constraints
```

Author: Done

*Take into consideration that the first label is used for the index creation.*

==== Approach

The connector allows to enforce the following constraints:

.Allowed configuration
|===
|config |values |default |description

|`schema.optimization.node.unique` |true/false |false | Create the `UNIQUE` constraint for the properties defined in the option `node.keys`
|`schema.optimization.node.key` |true/false |false | Create the `NODE KEY` constraint for the properties defined in the option `node.keys`
|`schema.optimization.relationship.unique` |true/false |false | Create the `UNIQUE` constraint for the properties defined in the option `relationship.keys`
|`schema.optimization.relationship.key` |true/false |false | Create the `RELATIONSHIP KEY` constraint for the properties defined in the option `relationship.keys`

> |`schema.type.constraint` |true/false |false | Creates the type constraints for node/relationships enforcing the type from the DataFrame schema

Reviewer 2: ```suggestion
|`schema.type.constraint` |true/false |false | Creates the type constraints for nodes/relationships enforcing the type from the DataFrame schema
```

Author: Done

*Take into consideration that the first label is used for the index creation.*

> ==== Approach

Reviewer: ```suggestion
==== With constraints
```

The connector allows to enforce the following constraints:

.Allowed configuration
|===
|config |values |default |description

|`schema.optimization.node` |UNIQUE/KEY/NONE |NONE | Create the `UNIQUE` or `NODE KEY` constraint for the properties defined in the option `node.keys`
|`schema.optimization.relationship` |UNIQUE/KEY/NONE |false | Create the `UNIQUE` or `NODE KEY` constraint for the properties defined in the option `relationship.keys`
|`schema.optimization` |TYPE/EXISTS/NONE |NONE | A comma separated list of values. Creates the type constraints for nodes/relationships enforcing the type and non-nullability from the DataFrame schema
|===

[NOTE]
Please consider that if you define more than one label we use the *first* for creating constraints.

===== Node constraints

> 
> ====== Using `schema.optimization.node`

Reviewer: ```suggestion
```

Author: Sorry why this?

Reviewer: For two reasons:

- It was too deeply nested.
- It was the only section header at that level.

**Enforcing unique constraint**

[NOTE]

> For a detailed description of how Neo4j handles unique constraints on nodes please go to the official https://neo4j.com/docs/cypher-manual/current/constraints/#unique-node-property[documentation]

Reviewer: ```suggestion
For a detailed description of how Neo4j handles unique constraints on nodes, see the https://neo4j.com/docs/cypher-manual/current/constraints/#unique-node-property[Cypher documentation].
```

Given the following example:

[source, scala]
----
    ds.write
      .format(classOf[DataSource].getName)
      .mode(SaveMode.Overwrite)
      .option("url", SparkConnectorScalaSuiteIT.server.getBoltUrl)
      .option("labels", ":Person:Customer")
      .option("node.keys", "surname")
      .option("schema.optimization.node", "UNIQUE")
      .save()
----

Under the hood the Spark connector will create the following constraint:

`CREATE CONSTRAINT `spark_NODE_UNIQUE-CONSTRAINT_Person_surname` IF NOT EXISTS FOR (e:Person) REQUIRE (e.surname) IS UNIQUE`

**Enforcing node key constraint**

[NOTE]

> For a detailed description of how Neo4j handles node key constraints please go to the official https://neo4j.com/docs/cypher-manual/current/constraints/#node-key[documentation]

Reviewer: ```suggestion
For a detailed description of how Neo4j handles node key constraints, see the https://neo4j.com/docs/cypher-manual/current/constraints/#node-key[Cypher documentation].
```

Given the following example:

[source, scala]
----
    ds.write
      .format(classOf[DataSource].getName)
      .mode(SaveMode.Overwrite)
      .option("url", SparkConnectorScalaSuiteIT.server.getBoltUrl)
      .option("labels", ":Person:Customer")
      .option("node.keys", "surname")
      .option("schema.optimization.node.keys", "KEY")
      .save()
----

Under the hood the Spark connector will create the following constraint:

`CREATE CONSTRAINT `spark_NODE_KEY-CONSTRAINT_Person_surname` IF NOT EXISTS FOR (e:Person) REQUIRE (e.surname) IS NODE KEY`

===== Relationship constraints

> 
> ====== Using `schema.optimization.relationship`

Reviewer: ```suggestion
```

**Enforcing unique constraint**

[NOTE]

> For a detailed description of how Neo4j handles unique constraints on relationships please go to the official https://neo4j.com/docs/cypher-manual/current/constraints/#unique-relationship-property[documentation]

Reviewer: ```suggestion
For a detailed description of how Neo4j handles unique constraints on relationships, see the https://neo4j.com/docs/cypher-manual/current/constraints/#unique-relationship-property[Cypher documentation].
```

Given the following example:

[source, scala]
----
    ds
      .write
      .mode(SaveMode.Overwrite)
      .format(classOf[DataSource].getName)
      .option("url", SparkConnectorScalaSuiteIT.server.getBoltUrl)
      .option("relationship", "MY_REL")
      .option("relationship.save.strategy", "keys")
      .option("relationship.source.labels", ":NodeA")
      .option("relationship.source.save.mode", "Overwrite")
      .option("relationship.source.node.keys", "idSource:id")
      .option("relationship.target.labels", ":NodeB")
      .option("relationship.target.node.keys", "idTarget:id")
      .option("relationship.target.save.mode", "Overwrite")
      .option("schema.optimization.relationship.keys", "UNIQUE")
      .option("relationship.keys", "foo,bar")
      .save()
----

Under the hood the Spark connector will create the following constraint:

`CREATE CONSTRAINT `spark_RELATIONSHIP_UNIQUE-CONSTRAINT_MY_REL_foo-bar` IF NOT EXISTS FOR ()-[e:MY_REL]->() REQUIRE (e.foo, e.bar) IS UNIQUE`

**Enforcing relationship key constraint**

[NOTE]

> For a detailed description of how Neo4j handles relationship key constraint please go to the official https://neo4j.com/docs/cypher-manual/current/constraints/#relationship-key[documentation]

Reviewer: ```suggestion
For a detailed description of how Neo4j handles relationship key constraint, see the https://neo4j.com/docs/cypher-manual/current/constraints/#relationship-key[Cypher documentation].
```

|===
|config |values |default |description

|`schema.optimization.node` |UNIQUE/KEY/NONE |NONE | Create the `UNIQUE` or `NODE KEY` constraint for the properties defined in the option `node.keys`
|`schema.optimization.relationship` |UNIQUE/KEY/NONE |false | Create the `UNIQUE` or `NODE KEY` constraint for the properties defined in the option `relationship.keys`

> |`schema.optimization` |TYPE/EXISTS/NONE |NONE | A comma separated list of values. Creates the type constraints for nodes/relationships enforcing the type from the DataFrame schema

Author: done

      .option("url", SparkConnectorScalaSuiteIT.server.getBoltUrl)
      .option("labels", ":Person:Customer")
      .option("node.keys", "surname")

>       .option("schema.optimization.node", "UNIQUE")

Reviewer 2: ```suggestion
      .option("schema.optimization.node.key", "UNIQUE")
```

Author: done

Reviewer 2: my bad, it's `.keys` and I don't see the change

      .option("url", SparkConnectorScalaSuiteIT.server.getBoltUrl)
      .option("labels", ":Person:Customer")
      .option("node.keys", "surname")

>       .option("schema.optimization.node", "KEY")

Reviewer 2: ```suggestion
      .option("schema.optimization.node.key", "KEY")
```

Author: done

Reviewer 2: my bad, it's `.keys` and I don't see the change

      .option("relationship.target.labels", ":NodeB")
      .option("relationship.target.node.keys", "idTarget:id")
      .option("relationship.target.save.mode", "Overwrite")

>       .option("schema.optimization", true)

Reviewer 2: should be `TYPE` and/or `EXISTS`?

Author: yes it's both
